certbot/certbot/_internal/cli/helpful.py

```diff
@@ -166,15 +160,45 @@ def _usage_string(self, plugins: Iterable[str], help_arg: Union[str, bool]) -> s
 
         return usage
 
-    def remove_config_file_domains_for_renewal(self, parsed_args: argparse.Namespace) -> None:
+    def remove_config_file_domains_for_renewal(self, config: NamespaceConfig) -> None:
         """Make "certbot renew" safe if domains are set in cli.ini."""
         # Works around https://github.com/certbot/certbot/issues/4096
-        if self.verb == "renew":
-            for source, flags in self.parser._source_to_settings.items(): # pylint: disable=protected-access
-                if source.startswith("config_file") and "domains" in flags:
-                    parsed_args.domains = _Default() if self.detect_defaults else []
+        if (config.argument_sources['domains'] == ArgumentSource.CONFIG_FILE and
+                self.verb == "renew"):
+            config.domains = []
+
+    def _build_sources_dict(self) -> Dict[str, ArgumentSource]:
+        # ConfigArgparse's get_source_to_settings_dict doesn't actually create
+        # default entries for each argument with a default value, omitting many
+        # args we'd otherwise care about. So in general, unless an argument was
+        # specified in a config file/environment variable/command line arg,
+        # consider it as having a "default" value
+        result = { action.dest: ArgumentSource.DEFAULT for action in self.actions }
+
+        for source_str, source_dict in self.parser.get_source_to_settings_dict().items():
+            if source_str.startswith('config_file'):
+                source = ArgumentSource.CONFIG_FILE
+            else:
+                source = {
+                    'defaults': ArgumentSource.DEFAULT,
+                    'env_var': ArgumentSource.ENV_VAR,
+                    'command_line': ArgumentSource.COMMAND_LINE
+                }[source_str]
+            if source == ArgumentSource.COMMAND_LINE:
+                (_, args) = source_dict['']
+                args = [arg for arg in args if arg.startswith('-')]
+                for arg in args:
+                    for action in self.actions:
+                        if arg in action.option_strings:
+                            result[action.dest] = source
+                            continue
+            else:
+                for arg, (action, _) in source_dict.items():
+                    result[action.dest] = source
```

Comment:
that's a good point -- i think we _will_ get undefined behavior based on whatever `items()` gives us last, but afaik an arg will never be listed in the `defaults` key if it's present elsewhere (https://github.com/bw2/ConfigArgParse/blob/master/configargparse.py#L633-L638). i'll add a test to see if we can make that behavior more well-defined!

Reply:
looking into it more, it looks like configargparse tries to avoid adding an argument source if one already exists. that said, they don't mention this in their docs, and they've got no unit tests to ensure that, so i added a test to track if that ever changes.

Reply:
ah, so they intentionally construct it so it only tells us about the most relevant source? that's very nice and would definitely solve this! I agree that having a test on our end to confirm is nice though, if that's not guaranteed. as for the test -- I agree it's nice to have a test to confirm this. I think the current test isn't quite there, since the ordering is still undefined and it might just have happened to work. instead, the test could just set it in both places and call `self.parser.get_source_to_settings_dict().items()` and check it's only in there once.

Reply:
what if instead we impose an ordering, i.e. iterate through the `get_source_to_settings_dict()` keys in a defined order? that way the test doesn't need to reach into `self.parser`, which is stateful and implementation-detaily